File: app/web/src/lib/aioPorts.js
// aioPorts.js — the browser-side mirror of app/aio_ports.go.
//
// The All-in-One form shows every instance's ports *before* a deploy exists, so
// the slot arithmetic has to run here too. This file and aio_ports.go must agree
// exactly; the Go side's TestAIOPortsNeverCollide / TestAIOPortsAvoidProductDefaults
// pin the invariants, and aioPorts.test-shape below documents the contract for
// anyone editing either half.
//
// If you change a family base, a slot offset or a member count here, change
// app/aio_ports.go in the same commit.

export const FAM = {
  MYSQL: 'mysql',
  PG: 'postgres',
  MONGO: 'mongodb',
  VALKEY: 'valkey',
  PROXY: 'proxysql',
  HAPROXY: 'haproxy',
  ORCH: 'orchestrator',
}

// Family port bases — deliberately far from every product default, so an
// instance can never collide with what the product would have chosen itself.
export const FAMILY_BASE = {
  [FAM.MYSQL]: 13000,
  [FAM.PG]: 15000,
  [FAM.PROXY]: 16000,
  [FAM.MONGO]: 17000,
  [FAM.HAPROXY]: 18000,
  [FAM.VALKEY]: 19000,
  [FAM.ORCH]: 20000,
}

export const SLOT_WIDTH = 10
export const SLOTS_PER_FAMILY = 100

// AIO_KINDS mirrors aioKinds in app/aio_ports.go: the catalog of features an
// All-in-One node can instantiate, in "Add feature" menu order.
//
// `supported` marks the families whose provisioner exists today. Unsupported
// kinds are still listed — greyed out with a reason — rather than hidden, so the
// menu shows where the node type is going instead of quietly omitting things.
export const AIO_KINDS = [
  { kind: 'ps', label: 'Percona Server', family: FAM.MYSQL, est: 700, supported: true },
  { kind: 'psrepl', label: 'PS Replication', family: FAM.MYSQL, cluster: true, min: 2, max: 5, def: 3, est: 700, supported: true },
  { kind: 'innodb', label: 'InnoDB Cluster / GR', family: FAM.MYSQL, cluster: true, min: 3, max: 9, def: 3, odd: true, est: 800, supported: true },
  { kind: 'pxc', label: 'PXC Cluster', family: FAM.MYSQL, cluster: true, min: 2, max: 5, def: 3, est: 900, supported: true },
  { kind: 'mysqlce', label: 'MySQL Community', family: FAM.MYSQL, est: 700, supported: true },
  { kind: 'mysqlcerepl', label: 'MySQL Replication', family: FAM.MYSQL, cluster: true, min: 2, max: 5, def: 3, est: 700, supported: true },
  { kind: 'mysqlceinnodb', label: 'MySQL InnoDB / GR', family: FAM.MYSQL, cluster: true, min: 3, max: 9, def: 3, odd: true, est: 800, supported: true },
  { kind: 'mariadb', label: 'MariaDB', family: FAM.MYSQL, est: 600, supported: true },
  { kind: 'mariadbrepl', label: 'MariaDB Replication', family: FAM.MYSQL, cluster: true, min: 2, max: 5, def: 3, est: 600, supported: true },
  { kind: 'mariadbgalera', label: 'MariaDB Galera', family: FAM.MYSQL, cluster: true, min: 3, max: 5, def: 3, odd: true, est: 800, supported: true },
  { kind: 'pg', label: 'PostgreSQL', family: FAM.PG, est: 300, supported: true },
  { kind: 'patroni', label: 'Patroni Cluster', family: FAM.PG, cluster: true, min: 2, max: 5, def: 3, est: 450, supported: true },
  { kind: 'repmgr', label: 'repmgr Cluster', family: FAM.PG, cluster: true, min: 2, max: 5, def: 3, est: 350, supported: true },
  { kind: 'spock', label: 'Spock Cluster', family: FAM.PG, cluster: true, min: 2, max: 5, def: 2, est: 350, supported: true },
  { kind: 'psmdb', label: 'PSMDB', family: FAM.MONGO, est: 500, supported: true },
  { kind: 'psmrs', label: 'PSMDB Replica Set', family: FAM.MONGO, cluster: true, min: 3, max: 7, def: 3, odd: true, est: 500, supported: true },
  { kind: 'psmdbsharded', label: 'PSMDB Sharded', family: FAM.MONGO, cluster: true, min: 5, max: 13, def: 5, est: 400, supported: true },
  { kind: 'valkey', label: 'Valkey', family: FAM.VALKEY, est: 120, supported: true },
  { kind: 'valkeycluster', label: 'Valkey Cluster', family: FAM.VALKEY, cluster: true, min: 3, max: 7, def: 3, est: 120, supported: true },
  { kind: 'proxysql', label: 'ProxySQL', family: FAM.PROXY, cluster: true, min: 1, max: 3, def: 1, est: 200, supported: true },
  { kind: 'haproxy', label: 'HAProxy', family: FAM.HAPROXY, est: 80, supported: true },
  { kind: 'orchestrator', label: 'Orchestrator', family: FAM.ORCH, est: 150, supported: true },
]

export const kindOf = (kind) => AIO_KINDS.find((k) => k.kind === kind)
export const familyOf = (kind) => kindOf(kind)?.family || ''

// Family display names for the grouped "Add feature" menu and version pickers.
export const FAMILY_LABEL = {
  [FAM.MYSQL]: 'MySQL',
  [FAM.PG]: 'PostgreSQL',
  [FAM.MONGO]: 'MongoDB',
  [FAM.VALKEY]: 'Valkey',
  [FAM.PROXY]: 'ProxySQL',
  [FAM.HAPROXY]: 'HAProxy',
  [FAM.ORCH]: 'Orchestrator',
}

// memberCount clamps a declared member count the way the Go planner does, so the
// port preview matches what will actually be provisioned.
export function memberCount(kind, members) {
  const k = kindOf(kind)
  if (!k) return 1
  if (!k.cluster) return 1
  if (!members || members < k.min) return k.def
  if (members > k.max) return k.max
  return members
}

// portsFor resolves one member's ports. Offsets must match aioPortsFor().
export function portsFor(kind, slot, member) {
  const fam = familyOf(kind)
  const base = (FAMILY_BASE[fam] ?? 0) + (slot + member) * SLOT_WIDTH
  const p = { base, client: base }
  switch (fam) {
    case FAM.MYSQL:
      p.admin = base + 1
      p.group = base + 2
      p.ist = base + 3
      p.sst = base + 4
      p.check = base + 5
      break
    case FAM.PG:
      p.rest = base + 1
      p.etcdCli = base + 2
      p.etcdPr = base + 3
      break
    case FAM.VALKEY:
      // Valkey's bus defaults to port+10000, which would leave the family's
      // range — it is pinned inside the slot instead (cluster-port).
      p.group = base + 1
      break
    case FAM.PROXY:
      p.admin = base + 1
      p.group = base + 2
      break
    case FAM.HAPROXY:
      p.check = base + 1
      p.admin = base + 2
      break
    default:
      break
  }
  return p
}

// portList is every port a member binds, sorted — the form's per-instance table.
export const portList = (p) =>
  [p.client, p.admin, p.group, p.ist, p.sst, p.check, p.rest, p.etcdCli, p.etcdPr]
    .filter((v) => v)
    .sort((a, b) => a - b)

// PORT_ROLE labels a port in the instance card's table.
export const PORT_ROLE = {
  client: 'client', admin: 'admin', group: 'cluster', ist: 'IST', sst: 'SST',
  check: 'check', rest: 'REST', etcdCli: 'etcd client', etcdPr: 'etcd peer',
}

// assignSlots walks instances in declaration order, giving each its first slot
// within its family. Positional and deterministic, so appending an instance
// never renumbers the ones before it — a deployed instance keeps its port.
export function assignSlots(instances) {
  const next = {}
  const out = {}
  for (const inst of instances) {
    const fam = familyOf(inst.kind)
    if (!fam) continue
    out[inst.id] = next[fam] || 0
    next[fam] = (next[fam] || 0) + memberCount(inst.kind, inst.members)
  }
  return out
}

// memberInst mirrors aioMemberInst: the per-member instance id.
export function memberInst(name, kind, member, total) {
  const base = sanitizeInst(name)
  const k = kindOf(kind)
  if (total <= 1 && !k?.cluster) return base
  return `${base}-n${member + 1}`
}

// sanitizeInst mirrors hostLabel() in app/dns.go — an instance name becomes a
// hostname, a directory and a systemd unit name, so it must survive all three.
export function sanitizeInst(s) {
  return String(s || '')
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-+|-+$/g, '')
}

// planMembers expands the declared instances into the concrete member list the
// form previews and the backend will provision.
export function planMembers(instances) {
  const slots = assignSlots(instances)
  const out = []
  for (const inst of instances) {
    const k = kindOf(inst.kind)
    if (!k) continue
    const total = memberCount(inst.kind, inst.members)
    for (let m = 0; m < total; m++) {
      out.push({
        inst: memberInst(inst.name, inst.kind, m, total),
        kind: inst.kind,
        family: k.family,
        group: k.cluster ? sanitizeInst(inst.name) : '',
        ports: portsFor(inst.kind, slots[inst.id], m),
      })
    }
  }
  return out
}

// --- MySQL flavor ----------------------------------------------------------
// Every one of these server packages declares Provides: mysql-server and conflicts
// with the others, so an All-in-One node has at most one MySQL flavor. It is
// derived from the instances, never chosen. Mirrors app/aio_ports.go.

export const FLAVOR_PS = 'ps'
export const FLAVOR_PXC = 'pxc'
export const FLAVOR_MARIADB = 'mariadb'
export const FLAVOR_MYSQLCE = 'mysqlce'

export const FLAVOR_LABEL = {
  [FLAVOR_PS]: 'Percona Server',
  [FLAVOR_PXC]: 'Percona XtraDB Cluster',
  [FLAVOR_MARIADB]: 'MariaDB',
  [FLAVOR_MYSQLCE]: 'MySQL Community',
}

export const flavorOfKind = (kind) => {
  if (kind === 'ps' || kind === 'psrepl' || kind === 'innodb') return FLAVOR_PS
  if (kind === 'pxc') return FLAVOR_PXC
  if (kind === 'mariadb' || kind === 'mariadbrepl' || kind === 'mariadbgalera') return FLAVOR_MARIADB
  if (kind === 'mysqlce' || kind === 'mysqlcerepl' || kind === 'mysqlceinnodb') return FLAVOR_MYSQLCE
  return ''
}

// A kind's shape is its topology, independent of whose build provides it.
export const SHAPE_SINGLE = 'single'
export const SHAPE_REPL = 'repl'
export const SHAPE_GR = 'gr'
export const SHAPE_GALERA = 'galera'

export const shapeOfKind = (kind) => {
  if (kind === 'ps' || kind === 'mariadb' || kind === 'mysqlce') return SHAPE_SINGLE
  if (kind === 'psrepl' || kind === 'mariadbrepl' || kind === 'mysqlcerepl') return SHAPE_REPL
  if (kind === 'innodb' || kind === 'mysqlceinnodb') return SHAPE_GR
  if (kind === 'pxc' || kind === 'mariadbgalera') return SHAPE_GALERA
  return ''
}

// mysqlFlavor returns { flavor, conflict, byFlavor } where byFlavor maps each
// flavor present to the instance names that pulled it in.
export function mysqlFlavor(instances) {
  const byFlavor = {}
  for (const i of instances) {
    const f = flavorOfKind(i.kind)
    if (!f) continue
    ;(byFlavor[f] ||= []).push(i.name)
  }
  const present = Object.keys(byFlavor)
  if (present.length > 1) return { flavor: '', conflict: true, byFlavor }
  return { flavor: present[0] || '', conflict: false, byFlavor }
}

// addBlockedReason explains why a kind cannot be added to the current node, or
// returns '' when it can. The menu shows the entry disabled with this text
// rather than hiding it, so the constraint is discoverable instead of mysterious.
export function addBlockedReason(kind, instances) {
  const k = kindOf(kind)
  if (!k) return 'Unknown feature'
  if (!k.supported) return `${k.label} instances are not implemented yet — use a dedicated ${k.label} node`
  const want = flavorOfKind(kind)
  if (!want) return ''
  const { byFlavor } = mysqlFlavor(instances)
  const other = Object.keys(byFlavor).filter((f) => f !== want)
  if (!other.length) return ''
  const names = other.map((f) => `${FLAVOR_LABEL[f]} (${byFlavor[f].join(', ')})`).join(' and ')
  return `${FLAVOR_LABEL[want]} can't share a container with ${names} — these server packages all provide mysql-server and conflict`
}

// estMemMB is the rough total footprint, for the form's sizing warning.
export const estMemMB = (instances) =>
  instances.reduce((t, i) => t + (kindOf(i.kind)?.est || 0) * memberCount(i.kind, i.members), 0)

// nextInstanceName picks the lowest unused "<prefix>NN" for a kind.
export function nextInstanceName(instances, kind) {
  const k = kindOf(kind)
  const prefix = k?.cluster ? `${kind}-cluster-` : kind
  const used = new Set(instances.map((i) => sanitizeInst(i.name)))
  for (let i = 1; ; i++) {
    const n = `${prefix}${String(i).padStart(2, '0')}`
    if (!used.has(n)) return n
  }
}

File: app/aio_ports.go
package main

import (
	"sort"
	"strconv"
	"strings"
)

// aio_ports.go — the All-in-One node's feature catalog and port allocator.
//
// An All-in-One node is ONE container running many database feature instances
// side by side. Two things follow from that and are decided here:
//
//   1. Nothing may listen on its product's default port. A stock Percona Server
//      wants 3306, a stock PostgreSQL 5432 — put two of either in one container
//      and the second fails to bind. Every instance therefore gets a private
//      *slot* of 10 consecutive ports, allocated per family, well away from the
//      defaults.
//   2. Which feature kinds exist, which family each belongs to, and how many
//      members each may have — the single source of truth the form, the
//      validator and the provisioners all read.
//
// The slot math is mirrored in web/src/lib/aioPorts.js so the designer can show
// the ports before a deploy exists. Change one, change the other; TestAIOPortsJS
// keeps them honest.

// ---------------------------------------------------------------- kinds

// AiO feature families. A family maps 1:1 onto "one package install serves every
// instance of it", which is why the version picker is per family, not per
// instance (see aioFamilyVersionShared).
const (
	famMySQL   = "mysql"
	famPG      = "postgres"
	famMongo   = "mongodb"
	famValkey  = "valkey"
	famProxy   = "proxysql"
	famHAProxy = "haproxy"
	famOrch    = "orchestrator"
)

// aioKind describes one feature a user can add to an All-in-One node.
type aioKind struct {
	Kind    string // the id stored in aioInstance.Kind
	Label   string // UI label
	Family  string // fam* above — decides packages, version sharing and port base
	Cluster bool   // true → the instance has Members>1 and gets one slot per member
	MinMem  int    // member count bounds (Cluster only)
	MaxMem  int
	DefMem  int
	OddOnly bool // member count must be odd (quorum-based topologies)
	// MemChoices, when set, is the ONLY member counts the kind accepts — a
	// fixed-topology kind, where a count in between would not describe a cluster
	// anyone would build. PSMDB Sharded is the case: its two counts are the two
	// setups the dedicated frame offers (see aioMongoShardedTopo).
	MemChoices []int
	// EstMemMB is a rough per-member resident footprint, used only to warn the
	// user before they ask one container to run thirty daemons.
	EstMemMB int
}

// aioKinds is the catalog. Order is the order the "Add feature" menu shows.
var aioKinds = []aioKind{
	{Kind: "ps", Label: "Percona Server", Family: famMySQL, EstMemMB: 700},
	{Kind: "psrepl", Label: "PS Replication", Family: famMySQL, Cluster: true, MinMem: 2, MaxMem: 5, DefMem: 3, EstMemMB: 700},
	{Kind: "innodb", Label: "InnoDB Cluster / GR", Family: famMySQL, Cluster: true, MinMem: 3, MaxMem: 9, DefMem: 3, OddOnly: true, EstMemMB: 800},
	{Kind: "pxc", Label: "PXC Cluster", Family: famMySQL, Cluster: true, MinMem: 2, MaxMem: 5, DefMem: 3, EstMemMB: 900},
	{Kind: "mysqlce", Label: "MySQL Community", Family: famMySQL, EstMemMB: 700},
	{Kind: "mysqlcerepl", Label: "MySQL Replication", Family: famMySQL, Cluster: true, MinMem: 2, MaxMem: 5, DefMem: 3, EstMemMB: 700},
	{Kind: "mysqlceinnodb", Label: "MySQL InnoDB / GR", Family: famMySQL, Cluster: true, MinMem: 3, MaxMem: 9, DefMem: 3, OddOnly: true, EstMemMB: 800},
	{Kind: "mariadb", Label: "MariaDB", Family: famMySQL, EstMemMB: 600},
	{Kind: "mariadbrepl", Label: "MariaDB Replication", Family: famMySQL, Cluster: true, MinMem: 2, MaxMem: 5, DefMem: 3, EstMemMB: 600},
	{Kind: "mariadbgalera", Label: "MariaDB Galera", Family: famMySQL, Cluster: true, MinMem: 3, MaxMem: 5, DefMem: 3, OddOnly: true, EstMemMB: 800},
	{Kind: "pg", Label: "PostgreSQL", Family: famPG, EstMemMB: 300},
	{Kind: "patroni", Label: "Patroni Cluster", Family: famPG, Cluster: true, MinMem: 2, MaxMem: 5, DefMem: 3, EstMemMB: 450},
	{Kind: "repmgr", Label: "repmgr Cluster", Family: famPG, Cluster: true, MinMem: 2, MaxMem: 5, DefMem: 3, EstMemMB: 350},
	{Kind: "spock", Label: "Spock Cluster", Family: famPG, Cluster: true, MinMem: 2, MaxMem: 5, DefMem: 2, EstMemMB: 350},
	{Kind: "psmdb", Label: "PSMDB", Family: famMongo, EstMemMB: 500},
	{Kind: "psmrs", Label: "PSMDB Replica Set", Family: famMongo, Cluster: true, MinMem: 3, MaxMem: 7, DefMem: 3, OddOnly: true, EstMemMB: 500},
	{Kind: "psmdbsharded", Label: "PSMDB Sharded", Family: famMongo, Cluster: true, MinMem: 5, MaxMem: 13, DefMem: 5, MemChoices: []int{5, 13}, EstMemMB: 400},
	{Kind: "valkey", Label: "Valkey", Family: famValkey, EstMemMB: 120},
	{Kind: "valkeycluster", Label: "Valkey Cluster", Family: famValkey, Cluster: true, MinMem: 3, MaxMem: 7, DefMem: 3, EstMemMB: 120},
	{Kind: "proxysql", Label: "ProxySQL", Family: famProxy, Cluster: true, MinMem: 1, MaxMem: 3, DefMem: 1, EstMemMB: 200},
	{Kind: "haproxy", Label: "HAProxy", Family: famHAProxy, EstMemMB: 80},
	{Kind: "orchestrator", Label: "Orchestrator", Family: famOrch, EstMemMB: 150},
}

var aioKindByID = func() map[string]aioKind {
	m := make(map[string]aioKind, len(aioKinds))
	for _, k := range aioKinds {
		m[k.Kind] = k
	}
	return m
}()

// aioKindOf looks up a kind; ok is false for an unknown kind id.
func aioKindOf(kind string) (aioKind, bool) {
	k, ok := aioKindByID[kind]
	return k, ok
}

// aioFamilyOf is the family a kind belongs to ("" when unknown).
func aioFamilyOf(kind string) string {
	if k, ok := aioKindByID[kind]; ok {
		return k.Family
	}
	return ""
}

// aioMemberCount is how many slots (daemons) an instance occupies, clamped to
// the kind's bounds. A non-cluster kind is always 1.
func aioMemberCount(kind string, members int) int {
	k, ok := aioKindByID[kind]
	if !ok {
		return 1
	}
	if !k.Cluster {
		return 1
	}
	if len(k.MemChoices) > 0 {
		// A fixed-topology kind snaps to the largest topology the count covers,
		// rather than clamping into a shape that does not exist. Snapping instead
		// of failing keeps a design saved before the choices narrowed deployable;
		// validateStack still flags the count so the user can correct it.
		best := 0
		for _, c := range k.MemChoices {
			if members >= c && c > best {
				best = c
			}
		}
		if best == 0 {
			return k.DefMem
		}
		return best
	}
	if members < k.MinMem {
		return k.DefMem
	}
	if members > k.MaxMem {
		return k.MaxMem
	}
	return members
}

// aioMemberChoiceOK reports whether a declared member count is one the kind
// accepts. Kinds without MemChoices accept anything inside their bounds, so the
// range/odd checks in validateStack remain the test for those.
func aioMemberChoiceOK(k aioKind, members int) bool {
	if len(k.MemChoices) == 0 {
		return true
	}
	for _, c := range k.MemChoices {
		if members == c {
			return true
		}
	}
	return false
}

// aioMemberChoicesText lists a kind's allowed member counts for a message —
// "5 or 13".
func aioMemberChoicesText(k aioKind) string {
	var parts []string
	for _, c := range k.MemChoices {
		parts = append(parts, strconv.Itoa(c))
	}
	switch len(parts) {
	case 0:
		return ""
	case 1:
		return parts[0]
	}
	return strings.Join(parts[:len(parts)-1], ", ") + " or " + parts[len(parts)-1]
}

// ---------------------------------------------------------------- MySQL flavor

// The MySQL family is the one place where feature kinds cannot share a container.
// Every one of these server packages declares Provides: mysql-server and conflicts
// with the others at the package level, so an AiO node has at most ONE MySQL
// flavor, derived from the instances present rather than chosen by the user.
const (
	flavorNone    = ""
	flavorPS      = "ps"      // percona-server-server
	flavorPXC     = "pxc"     // percona-xtradb-cluster-server
	flavorMariaDB = "mariadb" // MariaDB-server (mariadb.org)
	flavorMySQLCE = "mysqlce" // mysql-community-server (repo.mysql.com)
)

// A kind's *shape* is its topology, independent of whose build provides it: the
// three flavors that speak MySQL run the same standalone / replication / Group
// Replication code, and Galera is Galera whether PXC or MariaDB ships it. Keeping
// shape separate from flavor is what lets MySQL Community reuse the Percona paths
// outright — only the packages differ.
const (
	shapeNone   = ""
	shapeSingle = "single" // one standalone server
	shapeRepl   = "repl"   // primary + replicas, classic async/semi-sync
	shapeGR     = "gr"     // Group Replication (optionally InnoDB Cluster)
	shapeGalera = "galera" // Galera / wsrep
)

// aioMySQLFlavorOfKind is the flavor a single MySQL-family kind requires
// (flavorNone for kinds outside the family).
func aioMySQLFlavorOfKind(kind string) string {
	switch kind {
	case "ps", "psrepl", "innodb":
		return flavorPS
	case "pxc":
		return flavorPXC
	case "mariadb", "mariadbrepl", "mariadbgalera":
		return flavorMariaDB
	case "mysqlce", "mysqlcerepl", "mysqlceinnodb":
		return flavorMySQLCE
	}
	return flavorNone
}

// aioMySQLShape is a MySQL-family kind's topology (shapeNone outside the family).
func aioMySQLShape(kind string) string {
	switch kind {
	case "ps", "mariadb", "mysqlce":
		return shapeSingle
	case "psrepl", "mariadbrepl", "mysqlcerepl":
		return shapeRepl
	case "innodb", "mysqlceinnodb":
		return shapeGR
	case "pxc", "mariadbgalera":
		return shapeGalera
	}
	return shapeNone
}

// aioIsMariaDB reports whether a flavor speaks MariaDB's dialect rather than
// MySQL's. This is the only axis on which MariaDB genuinely differs here: its GTIDs
// are domain-server-seq, it attaches replicas with MASTER_USE_GTID, it has no SET
// PERSIST, and its client/daemon are mariadb/mariadbd.
func aioIsMariaDB(flavor string) bool { return flavor == flavorMariaDB }

// aioFlavorLabel names a flavor for progress lines and validation messages.
func aioFlavorLabel(flavor string) string {
	switch flavor {
	case flavorPS:
		return "Percona Server"
	case flavorPXC:
		return "Percona XtraDB Cluster"
	case flavorMariaDB:
		return "MariaDB"
	case flavorMySQLCE:
		return "MySQL Community"
	}
	return "MySQL"
}

// ---------------------------------------------------------------- ports

// Family port bases. Each instance member takes a 10-port slot starting at
// base + slotIndex*10, so no product ever sees its default port and the
// families cannot reach each other's ranges (aioSlotsPerFamily caps them).
var aioFamilyBase = map[string]int{
	famMySQL:   13000,
	famPG:      15000,
	famProxy:   16000,
	famMongo:   17000,
	famHAProxy: 18000,
	famValkey:  19000,
	famOrch:    20000,
}

// aioSlotWidth is how many ports one member owns. Everything a single daemon
// needs (client port, cluster transport, admin/REST, a co-located etcd) fits.
const aioSlotWidth = 10

// aioSlotsPerFamily bounds a family to 100 members (1000 ports) so ranges never
// collide. Far past what fits in one container; validation rejects earlier.
const aioSlotsPerFamily = 100

// aioPorts is one member's resolved port set. Zero means "not used by this kind".
type aioPorts struct {
	Base int `json:"base"` // slot base — also the primary/client port

	Client  int `json:"client"`            // mysqld 3306 / postgres 5432 / mongod 27017 / valkey 6379 / proxysql mysql iface
	Admin   int `json:"admin,omitempty"`   // mysqlx, proxysql admin, haproxy stats, orchestrator web
	Group   int `json:"group,omitempty"`   // galera gcomm / GR local address / valkey cluster bus / proxysql cluster iface
	IST     int `json:"ist,omitempty"`     // galera IST
	SST     int `json:"sst,omitempty"`     // galera SST
	Check   int `json:"check,omitempty"`   // clustercheck / haproxy read port
	REST    int `json:"rest,omitempty"`    // patroni REST API
	EtcdCli int `json:"etcdCli,omitempty"` // patroni's co-located etcd, client
	EtcdPr  int `json:"etcdPr,omitempty"`  // ...and peer
}

// aioPortsFor resolves the port set for member `member` (0-based) of an instance
// whose first slot is `slot`. Offsets inside a slot are fixed per family so a
// port's meaning is stable and readable in `aioctl ports`.
func aioPortsFor(kind string, slot, member int) aioPorts {
	fam := aioFamilyOf(kind)
	base := aioFamilyBase[fam] + (slot+member)*aioSlotWidth
	p := aioPorts{Base: base, Client: base}
	switch fam {
	case famMySQL:
		p.Admin = base + 1 // mysqlx
		p.Group = base + 2 // wsrep gcomm / group_replication_local_address
		p.IST = base + 3
		p.SST = base + 4
		p.Check = base + 5 // clustercheck
	case famPG:
		p.REST = base + 1 // patroni
		p.EtcdCli = base + 2
		p.EtcdPr = base + 3
	case famMongo:
		// mongod/mongos need one port each; the rest of the slot is headroom.
	case famValkey:
		// Valkey's cluster bus defaults to port+10000, which would land in
		// another family's range (19000+10000). Pin it inside the slot instead
		// and set cluster-port explicitly in the config.
		p.Group = base + 1
	case famProxy:
		p.Admin = base + 1
		p.Group = base + 2 // proxysql cluster interface
	case famHAProxy:
		p.Check = base + 1 // read/replica port
		p.Admin = base + 2 // stats
	case famOrch:
		// Orchestrator listens on one HTTP port (moved off its default :3000).
	}
	return p
}

// aioAllPorts lists every TCP port a member actually binds, for conflict checks
// and for the `aioctl ports` table.
func (p aioPorts) list() []int {
	var out []int
	for _, v := range []int{p.Client, p.Admin, p.Group, p.IST, p.SST, p.Check, p.REST, p.EtcdCli, p.EtcdPr} {
		if v != 0 {
			out = append(out, v)
		}
	}
	sort.Ints(out)
	return out
}

// aioAssignSlots walks a node's instances in order and assigns each one its
// first slot index within its family, returning the slots keyed by instance id.
// Allocation is positional and deterministic: the same design always produces
// the same ports, so a redeploy never moves a running instance's port. A
// cluster instance consumes one slot per member.
func aioAssignSlots(instances []aioInstance) map[string]int {
	next := map[string]int{}
	out := make(map[string]int, len(instances))
	for _, in := range instances {
		fam := aioFamilyOf(in.Kind)
		if fam == "" {
			continue
		}
		out[in.ID] = next[fam]
		next[fam] += aioMemberCount(in.Kind, in.Members)
	}
	return out
}

// aioSlotsUsed is how many slots a family consumes for a node's instances —
// used by validation to reject a design that would overflow its range.
func aioSlotsUsed(instances []aioInstance, family string) int {
	n := 0
	for _, in := range instances {
		if aioFamilyOf(in.Kind) == family {
			n += aioMemberCount(in.Kind, in.Members)
		}
	}
	return n
}

// ---------------------------------------------------------------- web endpoints

// aioWebEndpoint is an HTTP interface an instance serves: which of its ports, and
// what to call it.
type aioWebEndpoint struct {
	Label string `json:"label"`
	Port  int    `json:"port"`
	Path  string `json:"path"`
}

// aioWebEndpoints lists the HTTP interfaces one member serves.
//
// These exist so the manager can offer a real link instead of a host:port string a
// user has to assemble by hand — and so aioCreateContainer knows to publish them.
// Only the client port was ever published, which left HAProxy's stats page and
// Patroni's REST API unreachable from the host even with the export ticked.
func aioWebEndpoints(kind string, p aioPorts) []aioWebEndpoint {
	switch kind {
	case "orchestrator":
		// Orchestrator's web UI and its API are the same listener.
		return []aioWebEndpoint{{Label: "Orchestrator", Port: p.Client, Path: "/"}}
	case "haproxy":
		return []aioWebEndpoint{{Label: "HAProxy stats", Port: p.Admin, Path: "/stats"}}
	case "patroni":
		// JSON rather than a UI, but it is how you read cluster state over HTTP.
		return []aioWebEndpoint{{Label: "Patroni REST", Port: p.REST, Path: "/cluster"}}
	}
	return nil
}

// aioPublishPorts is every container port an instance should publish when it opts
// into an export: the client port, plus any HTTP endpoint that is on a different
// port (a web UI is useless if only the client port is mapped).
func aioPublishPorts(kind string, p aioPorts) []int {
	out := []int{p.Client}
	seen := map[int]bool{p.Client: true}
	for _, w := range aioWebEndpoints(kind, p) {
		if w.Port > 0 && !seen[w.Port] {
			out = append(out, w.Port)
			seen[w.Port] = true
		}
	}
	return out
}
